### ISSUE TITLE
Blank page after merge.

### ISSUE THREAD

**Issue by morkai** [py-pdf/pypdf#2260]
I'm trying to add a short text to each page of an existing document. The input document I'm using for testing has 18 pages and after running it through the specified script two pages containing only text are blank.

I've got permission to share the first three pages of the document (one of the failing pages is the page number 3).

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
Windows-10-10.0.22621-SP0

$ python -c "import pypdf;print(pypdf._debug_versions)"
pypdf==3.16.4, crypt_provider=('pycryptodome', '3.10.1'), PIL=9.4.0
```

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
from pypdf import PdfWriter, PdfReader, Transformation

input_pdf = PdfReader(open("blank_merge.input.pdf", "rb"))
output_pdf = PdfWriter()

for page_no, input_page in enumerate(input_pdf.pages):
    print(page_no + 1)
    footer_pdf = PdfReader(open("blank_merge.footer.pdf", "rb"))
    footer_page = footer_pdf.pages[0]
    footer_page.add_transformation(Transformation().rotate(0).translate(tx=0, ty=0))
    input_page.merge_page(footer_page)
    output_pdf.add_page(input_page)

outputStream = open("blank_merge.output.pdf", "wb")
output_pdf.write(outputStream)
outputStream.close()
```

In the above script we are merging the `footer_page` into the `input_page`, but if we swap the order and merge the `input_page` into the `footer_page` everything works (but is noticably slower):

```python
from pypdf import PdfWriter, PdfReader, Transformation

input_pdf = PdfReader(open("blank_merge.input.pdf", "rb"))
output_pdf = PdfWriter()

for page_no, input_page in enumerate(input_pdf.pages):
    print(page_no + 1)
    footer_pdf = PdfReader(open("blank_merge.footer.pdf", "rb"))
    footer_page = footer_pdf.pages[0]
    footer_page.add_transformation(Transformation().rotate(0).translate(tx=0, ty=0))
    footer_page.merge_page(input_page)
    output_pdf.add_page(footer_page)

outputStream = open("blank_merge.output.pdf", "wb")
output_pdf.write(outputStream)
outputStream.close()
```

[blank_merge.input.pdf](https://github.com/py-pdf/pypdf/files/13040301/blank_merge.input.pdf) - the input PDF - notice that the 3rd page is all text
[blank_merge.footer.pdf](https://github.com/py-pdf/pypdf/files/13040312/blank_merge.footer.pdf) - this is created through Chrome's 'Print to PDF' feature (PDF created with the reportlab package also doesn't work)
[blank_merge.output.pdf](https://github.com/py-pdf/pypdf/files/13040303/blank_merge.output.pdf) - the result of merging the footer into the input - notice that the 3rd page is blank
[blank_merge.output_reversed.pdf](https://github.com/py-pdf/pypdf/files/13040498/blank_merge.output_reversed.pdf) - the result of merging the input into the footer - notice that the 3rd page has the original contents and the extra footer

**Comment by stefan6419846** [comment]
I am not completely sure why, but the following approach seems to work (id est using the `clone_from` version):

```python3
from pypdf import PdfReader, PdfWriter, Transformation


writer = PdfWriter(clone_from="blank_merge.input.pdf")

for page_no, page in enumerate(writer.pages):
    print(page_no + 1)
    footer_pdf = PdfReader("blank_merge.footer.pdf")
    footer_page = footer_pdf.pages[0]
    footer_page.add_transformation(Transformation().rotate(0).translate(tx=0, ty=0))
    page.merge_page(footer_page)

writer.write("blank_merge.output.pdf")
```

**Comment by bigatti** [comment]
@morkai Could you test the same PDF, but using pypdf version 3.10.0 ?

I have the same problem, after version 3.10.0 there is a problem.

[Compare 3.10.0 to 3.11.0](https://github.com/py-pdf/pypdf/compare/3.10.0...3.11.0)

**Comment by morkai** [comment]
@bigatti Yes, it works in 3.10.0 and breaks from 3.11.0.

**Comment by stefan6419846** [comment]
Still no solution, but it seems like omitting the second page avoids the final blank page.

The offending section is https://github.com/py-pdf/pypdf/blob/4f2cd3439c6f074c515fe347ef92ba0bb44a9e37/pypdf/_page.py#L1122-L1126 This generates a standalone page object (`84 0 obj`) which seems to have the correct content (id est the content of the original page 3), but is never referenced. The actually referenced page content is in xref 79, but cannot be decoded as it uses plaintext while declaring the `/FlateDecode` filter.

Generated broken file (without the irrelevant first page): [result_3.11.0.pdf](https://github.com/user-attachments/files/18208252/result_3.11.0.pdf)

**Comment by jonashaag** [comment]
I'm interested in helping debug this. What's the best way to do so for someone completely unfamiliar with the codebase?

**Comment by stefan6419846** [comment]
Thanks for your interest. I would recommend starting with the method referenced in my previous comment. Where is the not referenced page object 84 generated and why is the broken object 79 used here? Determining why the filter is not applied to the page on writing might be more complex.

**Comment by stefan6419846** [comment]
This seems to indicate that while the data is there, something prevents rendering them - maybe because of broken encoding or wrong cross-references (as seen previously), maybe due to another issue. What the issue is and where it originates from is what needs further analysis and is currently unclear for me.

**Comment by jonashaag** [comment]
My issue seems not identical, since I can't reproduce it with any version down to 3.7 (did not test further). Interestingly there are a bunch of JBIG2Decode errors so maybe it's because of missing support.

**Comment by stefan6419846** [comment]
According to the specification, JBIG2 should only affect images and not texts etc. You might want to test with #3163.

> My issue seems not identical, since I can't reproduce it with any version down to 3.7

Am I understanding it correctly that your issue is related to a recent change? In general, knowing the original commit which changed the behavior might help with debugging to better identify any undesired regression.

**Comment by larsga** [comment]
If this truly is the same issue as #3236 I have a little clue: it works on MacOS, but it fails on Windows. This is 100% consistent: always works on MacOS, never works on Windows.

**Comment by stefan6419846** [comment]
The same issue has been discovered in https://github.com/py-pdf/fpdf2/issues/1570#issuecomment-3278710350, but I unfortunately still have no clue where this fails.

**Comment by rhn** [comment]
I hit what seems like the same issue and brought it down to a minimal test case: https://github.com/py-pdf/pypdf/issues/3568

The files are <20KiB in size, so in the range of manual inspection. When I tried, though, I couldn't get useful debugging output on the internal objects. Objects typically print as {}, which is less than helpful. Any hints?

**Comment by stefan6419846** [comment]
> Objects typically print as {}, which is less than helpful. Any hints?

If you mean the page content stream as the object, this might indeed be correct. To get the corresponding data, call `content_stream.get_data()`, altough this might lead to unexpected side effects or issues given something is wrong about the encoding handling.

**Comment by rhn** [comment]
Yes, I wanted to get an idea what data is carried inside objects while they are getting passed around, but I couldn't easily find the data through all the inheritance. Weren't references stored as this kind of objects, separately from content?

 When printing get_data(), the objects themselves looked similar, except the broken one had been encapsulated within a single-letter call - I couldn't find what that means in the spec, either.

**Comment by stefan6419846** [comment]
> Weren't references stored as this kind of objects, separately from content?

Please be more specific what you are referencing to. There are some object references (`IndirectObject`s), but they are not necessarily independent from the content.

> except the broken one had been encapsulated within a single-letter call - I couldn't find what that means in the spec, either.

Please be more specific on what you mean. What single-letter calls?

**Comment by stefan6419846** [comment]
> > except the broken one had been encapsulated within a single-letter call - I couldn't find what that means in the spec, either.
> 
> Please be more specific on what you mean. What single-letter calls?

Do you mean this?

```
23 0 obj
<<
/Filter /FlateDecode
/Length 250
>>
stream
q
/Span <<
  /MCID 0
>> BDC
q
1 0 0 -1 70.86614 761.09064 cm
/c0 cs
0 scn
BT
0 Tr
/f0 15.4 Tf
1 0 0 -1 0 0 Tm
[(\000\001\000\002\000\003\000\003\000\002\000\004\000\005\000\006\000\003\000\007) -7 (\000\b\000\t\000\002\000\n\000\004)] TJ
ET
Q
EMC
Q


endstream
endobj
```

This is indeed the faulty stream (it should use the *FlateDecode* filter as indicated by the header data, but is unencoded plain text). Then encapsulating `q ... Q` are completely valid and make sure that the graphics state is isolated correctly.

My short code:

```python3
from io import BytesIO

from pypdf import PageObject, PdfReader, PdfWriter


reader = PdfReader('test-1.pdf')
writer = PdfWriter()

lhs = reader.get_page(3)
writer.add_page(lhs)

lhs = reader.get_page(1)
lhs.merge_page(PageObject.create_blank_page(reader))
writer.add_page(lhs)


buffer = BytesIO()
writer.write(buffer)

print(buffer.getvalue().decode('latin1'))
```

**Comment by rhn** [comment]
Yes, I think it was `q`/`Q`, as seen using get_data before/after merging the empty page. Sorry, I was debugging this over a week ago and I didn't remember exactly. I'm glad that you managed to reconstruct it with my vague report.

**Comment by stefan6419846** [comment]
I think I am slowly beginning to understand the underlying issue. As previously mentioned, it appears to originate from https://github.com/py-pdf/pypdf/blob/4f2cd3439c6f074c515fe347ef92ba0bb44a9e37/pypdf/_page.py#L1114-L1126

It seems that easy cases can be solved by simply `.indirect_reference` to the end of line 1126 - leading to some test failures for functionality unrelated at first sight. However, this is not the actual issue. Instead, the *AttributeError* in line 1122 complains:

> 'PdfReader' object has no attribute '_objects'

The comment for the fallback seems to be wrong in some way. Attempting to modify the reader would also be incorrect, as it should be read-only. Therefore, my previous suggestion of avoiding the reader completely and using `PdfWriter(clone_from=...)` still works and is probably the most reliable solution.

The only reliable and future-proof solution I can see for now is to deprecate the call to `merge_page()` (and especially `replace_contents()`) for pages that are not attached to a *PdfWriter* instance. Messing with pages that are only attached to a *PdfReader*, which should be read-only by design, is a bad approach.

I am aware that this restriction would require adapting code that has not previously failed, but which has always exposed the risk of ending up with blank pages.

If there are any objections to this or any viable alternative solutions, please comment accordingly.

**Comment by rhn** [comment]
I took the advice and rewrote my code to use the writer instead of the reader. 

That means I create an empty page in the reader and I merge my two source pages into it, instead of merging my source pages together. It's a small annoyance: now I have to explicitly declare page size, but I don't know if I should use media_box or which value for the create_blank_page argument.

The other annoyance is that now I need to carry the output document around. It makes managing multiple outputs less straightforward: I ended up passing a closure rather then a document, but I imagine coming up with stuff like that might be difficult for less experienced coders.

**Comment by stefan6419846** [comment]
> That means I create an empty page in the reader and I merge my two source pages into it, instead of merging my source pages together. It's a small annoyance: now I have to explicitly declare page size, but I don't know if I should use media_box or which value for the create_blank_page argument.

It should not be required to do larger changes to this code - it should be sufficient to just call `add_page` first and use its returned page for further actions, which are in-place operations:

```python3
from io import BytesIO

from pypdf import PageObject, PdfReader, PdfWriter


reader = PdfReader("test-1.pdf")
writer = PdfWriter()

lhs = reader.get_page(3)
writer.add_page(lhs)

lhs = reader.get_page(1)
lhs_writer = writer.add_page(lhs)
lhs_writer.merge_page(PageObject.create_blank_page(reader))

writer.write("output.pdf")
```

> The other annoyance is that now I need to carry the output document around. It makes managing multiple outputs less straightforward: I ended up passing a closure rather then a document, but I imagine coming up with stuff like that might be difficult for less experienced coders.

Without any real example code, this is hard to evaluate.
